app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt

```diff
@@ -17,60 +17,100 @@
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import java.io.IOException
+
+// GitHub page API is 1 based: https://developer.github.com/v3/#pagination
+private const val GITHUB_STARTING_PAGE_INDEX = 1
 
 /**
  * Repository class that works with local and remote data sources.
  */
-class GithubRepository(
-    private val service: GithubService,
-    private val cache: GithubLocalCache
-) {
+@FlowPreview
+@ExperimentalCoroutinesApi
+class GithubRepository(private val service: GithubService) {
 
-    // keep the last requested page. When the request is successful, increment the page number.
-    private var lastRequestedPage = 1
+    // keep the list of all results received
+    private val inMemoryCache = mutableListOf<Repo>()
+
+    // keep channel of results. The channel allows us to broadcast updates so
+    // the subscriber will have the latest data
+    private val searchResults = ConflatedBroadcastChannel<RepoSearchResult>()
 
-    // LiveData of network errors.
-    private val networkErrors = MutableLiveData<String>()
+    // keep the last requested page. When the request is successful, increment the page number.
+    private var lastRequestedPage = GITHUB_STARTING_PAGE_INDEX
 
     // avoid triggering multiple requests in the same time
     private var isRequestInProgress = false
 
     /**
-     * Search repositories whose names match the query.
+     * Search repositories whose names match the query, exposed as a stream of data that will emit
+     * every time we get more data from the network.
      */
-    fun search(query: String): RepoSearchResult {
+    suspend fun getSearchResultStream(query: String): Flow<RepoSearchResult> {
         Log.d("GithubRepository", "New query: $query")
         lastRequestedPage = 1
         requestAndSaveData(query)
 
-        // Get data from the local cache
-        val data = cache.reposByName(query)
-
-        return RepoSearchResult(data, networkErrors)
+        return searchResults.asFlow()
     }
 
-    fun requestMore(query: String) {
+    suspend fun requestMore(query: String) {
+        if (isRequestInProgress) return
         requestAndSaveData(query)
+        lastRequestedPage++
     }
 
-    private fun requestAndSaveData(query: String) {
+    suspend fun retry(query: String) {
         if (isRequestInProgress) return
+        requestAndSaveData(query)
+    }
 
+    private suspend fun requestAndSaveData(query: String) {
         isRequestInProgress = true
-        searchRepos(service, query, lastRequestedPage, NETWORK_PAGE_SIZE, { repos ->
-            cache.insert(repos) {
-                lastRequestedPage++
-                isRequestInProgress = false
+
+        val apiQuery = query + IN_QUALIFIER
+        try {
+            val response = service.searchRepos(apiQuery, lastRequestedPage, NETWORK_PAGE_SIZE)
+            Log.d("GithubRepository", "response $response")
```

Comment:
Was this log statement left here accidentally?

Reply:
May be useful when playing with the code

Reply:
We do leave Logs in code so it's easier for the developers to see what happens at different points in the app. 
I'll leave these here for now